### PR TITLE
proposal: introduce join reorder by dp subset

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Add proposal document for join reorder by dp-subset.

related: #8025 

### Check List 

Tests 

 - No code

### CODE REVIEW

File: docs/design/2018-10-20-join-reorder-dp-v1.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+There’s no compatibility issue.
+
+## Implementation
+
+First, extract the join nodes. If the number of nodes is less than x, we go into DP-SUB.
+
+For DP-SUB:
+
+1. Enlarge its edges by constant propagation.
```
Comment (eurekaka) [inline docs/design/2018-10-20-join-reorder-dp-v1.md]: If we do not consider `OtherCondition`, then constant propagation should remove edges from the join graph? e.g, `t1.a = t2.a and t2.a = 1` is converted to `t1.a = 1 and t2.a = 1`.

Comment (winoros) [reply docs/design/2018-10-20-join-reorder-dp-v1.md]: One side condition will be pushed down to join already.
We don't need to hold them when do join reorder.

Comment (eurekaka) [reply docs/design/2018-10-20-join-reorder-dp-v1.md]: Yes, I mean `Enlarge its edges by constant propagation` should be `Remove its edges by constant propagation`?

Comment (winoros) [reply docs/design/2018-10-20-join-reorder-dp-v1.md]: Oh i got it.
This propagation mentioned here is mainly to solve the following case:
```
t1.a=t2.a and t2.a=t3.a
```
There's actually one edge `t1.a=t3.a` and may be a edge involved in best choice.

And back to your case, that convention eats one edge and may cause the final plan worse than we expected. We need to come up with a way to handle it.

**Comment by zz-jason** [comment]
@lamxTyler @eurekaka @XuHuaiyu PTAL

**Pull request merged**